### DISCUSSION TITLE
Feature Sliced Review 🍰

### DISCUSSION THREAD

**Discussion by azinit** [EliseyMartynov/fs-vue#2] [General]
**Во-первых, спасибо большое что смог заняться этим! 🚀 🚀** 

Отпишу мысли тут, потом бы в GH-Discussions трансформировать, [как тут](https://github.com/yesnoruly/loripsum-generator/issues/7)

> В упомянутой ишьюс кста тож отписал, почему формат PR тут для ревью не оч подходит

---

P.S. 
- [Ссыль на воркспейс с кодом для ревью](https://github.dev/EliseyMartynov/fs-vue)
- [Ссыль на "оригинальную" TodoApp на реакте](https://codesandbox.io/s/github/feature-sliced/examples/tree/master/todo-app)

**Comment by azinit** [comment]
## Начну, с хорошего)

[`praise:`](https://conventionalcomments.org/) Здорово, что удалось переиспользовать shared-модули "as-is", без всякого шаманства

![image](https://user-images.githubusercontent.com/42924400/143136362-b98e529f-a7e2-4fa0-a1d6-45e217730d37.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Получилось довольно таки соответствующе FS по структуре и конвенциям)

Посмотрим что потом @niyazm524 скажет, но кажется прям топчик вышло (даже если чисто внешне на структуру смотреть)

![image](https://user-images.githubusercontent.com/42924400/143136534-70d1f96c-c488-4313-b927-2fdb491c1df5.png)

**Reply by niyazm524** [reply]
Да, согласен, + не догадался бы компонент RouterView в pages оставить)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) С провайдерами тоже довольно интересно вышло)

Прям по канону

![image](https://user-images.githubusercontent.com/42924400/143136827-01f1cb73-bc01-49d2-b983-866a74740e5d.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Хз что @niyazm524 скажет, но кажется довольно интересно получилось с Public API компонентов разобраться 👍 

![image](https://user-images.githubusercontent.com/42924400/143137023-137bc623-d71e-48d7-88e8-e5e180a5314d.png)

![image](https://user-images.githubusercontent.com/42924400/143137116-f7b8fbfc-4b4a-4956-a154-e5f49cef2823.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Круто также, что удалось передать мысль о "композиции логики"

![image](https://user-images.githubusercontent.com/42924400/143137227-91acc345-803c-49e4-90b7-e259c6267bf4.png)

![image](https://user-images.githubusercontent.com/42924400/143137242-10e177cf-42cc-4860-ba3b-bc619845d0dc.png)

![image](https://user-images.githubusercontent.com/42924400/143137290-727c21af-22f2-4f11-8145-63fff9ccb5a1.png)

**Comment by azinit** [comment]
## Теперь криты

Их нет, ни за что существенного глаз не зацепился

~~Аж свербит :D~~

**Comment by azinit** [comment]
## Теперь прочие замечания

[`nitpick(non-blocking):`](https://conventionalcomments.org/) Убрать бы все глобальные стили на уровень app

![image](https://user-images.githubusercontent.com/42924400/143139087-e4e52e33-3897-4ac7-9d19-6164f53fcfb9.png)


[`thoughts(non-blocking):`](https://conventionalcomments.org/) Похожий недочет был и в оригинальном TodoApp
Но лучше чтоб все глобальные стили в одном месте декларировались

https://github.com/feature-sliced/examples/commit/a20375a1296386012aba1176d8af1eff8961fd5a

**Comment by azinit** [comment]
[`nitpick(non-blocking):`](https://conventionalcomments.org/) Сейм, это подключение глобальных стилей - желательно вынести на уровень _src/app/styles/index.scss_ как в [оригинале](https://user-images.githubusercontent.com/42924400/143139285-5d46c763-7b20-4bd9-b42c-e51ed8abdd56.png)

![image](https://user-images.githubusercontent.com/42924400/143139211-8b54facc-12c3-404e-ba3d-2a792091c2e9.png)

[`thoughts:`](https://conventionalcomments.org/) Если импорты нод-модульных стилей подтупливают, мб этот коммит поможет 🤔 
https://github.com/feature-sliced/examples/commit/6a83c96666092b1339e70864d18175bbecbd72da

**Comment by azinit** [comment]
[`issue(non-blocking):`](https://conventionalcomments.org/) Хз насколько во vuex приемлемы концепции селекторов, но прям не оч, что при использовании приходится эту логику дублировать

![image](https://user-images.githubusercontent.com/42924400/143139654-f3ec35bb-236c-40e4-91cf-1d753a38075e.png)

[`suggestion(non-blocking):`](https://conventionalcomments.org/) По-хорошему сделать так, чтобы эта логика определялась
на уровне taskModel, а логика страницы лишь композировала бы все это и оставалась тонкой при этом (как в [оригинале](https://user-images.githubusercontent.com/42924400/143139592-f4b13287-34bb-41bf-82b3-d67190b2b772.png))

**Reply by azinit** [reply]
@niyazm524 мб ты тут что подскажешь

**Comment by azinit** [comment]
[`nitpick(non-blocking):`](https://conventionalcomments.org/) Вот прям вообще некрит - но если бы получилось эту логику модели данных вынести в отдельный `model.ts` файл, где велась бы работа с vuex / событиями и прочим - прям было бы топчик

![image](https://user-images.githubusercontent.com/42924400/143139814-fd1570b9-7b80-44b9-bc37-72ad2f7b494e.png)

[`thoughts(non-blocking):`](https://conventionalcomments.org/) Но мы [у себя в TodoApp](https://user-images.githubusercontent.com/42924400/143139920-23287ac8-e629-4e32-97d0-bb2e68b2ccb8.png) так не делали, так что рили некрит)


UPD: Мб @niyazm524 что подскажет

**Comment by azinit** [comment]
[`thoughts(non-blocking):`](https://conventionalcomments.org/) Поначалу показалось "Ух жирная моделька, прям нарезать надо"

Но потом вчитавшись стало ясно - что там все то минимально необходимое + с обвязками вокруг vuex

Мб @niyazm524 подскажет как можно упростить модельку сущности

![image](https://user-images.githubusercontent.com/42924400/143140251-21c7a582-268a-4cc3-946d-931c2cc9df49.png)

**Comment by azinit** [comment]
[`thoughts(non-blocking):`](https://conventionalcomments.org/) Кста хз насколько так принято в мире vue
Но это изначальное деление по "сабмоделям" с реэкспортом ключевых из effector изначально пошло)

Т.е. мб это даже избыточно, и хватило бы обычных реэкспортов

**А с другой стороны**, классно что под это удалось подвязать префиксацию нейминга :+1:

> @niyazm524, опять же, поправь если ошибаемся

![image](https://user-images.githubusercontent.com/42924400/143140371-935201fd-99db-4306-b8c6-e1bd7d1adef0.png)

**Comment by azinit** [comment]
[`thoughts(non-blocking):`](https://conventionalcomments.org/) В целом ощущение, что в страницах все равно слишком много логики остается

Возможно из-за того что "сливаются" импорты и отдельно сетапинг юнитов для темплейта

![image](https://user-images.githubusercontent.com/42924400/143141030-9200ece8-7914-46c5-8b23-8f485a2079c0.png)

**Comment by azinit** [comment]
[`question(non-blocking):`](https://conventionalcomments.org/) Чутка смутил момент с неймспейсами
У нас тип vuex тоже глобальный стор, где напрямую к каждому "министору" надо обращаться по имени?)
(а не по экспортируемому юниту, как в эффекторе)

[`thoughts(non-blocking):`](https://conventionalcomments.org/) Энивей, кажется, что такое все равно как будто можно решать геттерами/селекторами но мб глаз замылен и это ОК на самом деле выглядит

![image](https://user-images.githubusercontent.com/42924400/143141111-2c498d2f-42e6-4b2b-af7a-37c87e954933.png)

**Reply by niyazm524** [reply]
Для Options API были придуманы хелперы типа mapGetters, но в Vuex 4 это считается нормой

**Comment by azinit** [comment]
## И потихоньку вопросы по Vue

[`question:`](https://conventionalcomments.org/) @niyazm524 ОК ли делать так композицию разметки и логики в мире Vue? Или можно лучше/проще/по-другому?

![image](https://user-images.githubusercontent.com/42924400/143141687-cff0f1ce-742b-46c2-91e2-c56719a16d5c.png)

**Reply by niyazm524** [reply]
Можно конечно эту логику вынести в отдельный компонент с двумя слотами / одним слотом и пропсами, но вообще, если это больше нигде не обрабатывается, не критично.

Другое дело, что лучше было бы сделать что-то вроде хука, при вызове функции которой делался бы `router.replace` на страницу `{ name: 'Error404', path: route.fullPath }`, насколько помню, так должно работать, и весьма незаметно для пользователя.

**Reply by EliseyMartynov** [reply]
@niyazm524 так было в "оригинальном" ToDo example у FS. Поэтому так и написано. А так да - это классика делать error page :)

**Comment by azinit** [comment]
[`question:`](https://conventionalcomments.org/) @niyazm524 Знаешь как проблему с типами полечить?

![image](https://user-images.githubusercontent.com/42924400/143141703-9f4abd88-a976-44ce-b3ea-d886c37e0c9a.png)

**Reply by niyazm524** [reply]
Я ещё не работал с script setup, на прод такое не допустят пока :(

**Reply by azinit** [reply]
> на прод такое не допустят пока :(

@niyazm524 Т.е. это далеко от RealWorldVueProject на данный момент?)

**Reply by niyazm524** [reply]
Не то чтобы критично, просто некоторое API не обкатано ещё, а так, как пример архитектуры, думаю, весьма хорошо. К RealWorldVueProject приблизило бы использование error-pages. А так, даже использование Vite это уже ред флаг для коммерческого проекта, по крайней мере пока

**Comment by azinit** [comment]
[`thoughts:`](https://conventionalcomments.org/) Прям флешбеки от redux-toolkit...

![image](https://user-images.githubusercontent.com/42924400/143141778-8838a4d4-7132-4604-86e0-b0533c776c2d.png)

**Reply by niyazm524** [reply]
Когнитивная сложность _/‾

**Reply by azinit** [reply]
@niyazm524 А как упростить?

**Reply by azinit** [reply]
UPD: А, увидел ниже - т.е. предлагаешь полагаясь на те статьи модель данных переписать? (СТМ обработку)

**Reply by niyazm524** [reply]
На самом деле не призываю, мне кажется, это больше вкусовщина — кому как нравится
Но предлагать — да, предлагаю)

**Comment by azinit** [comment]
## Итого

В целом кажется все оч даже хорошо, все вещи что отписал - скорее ToImprove (даже критов не нашлось)

Остались вопросы разве что по специфике Vue, но думаю @niyazm524 ответит

**Comment by azinit** [comment]
Дождемся вердикта от @niyazm524, влития правок - и думаю можно будет организовывать сэндбокс и добавление в общий [список примеров)](https://feature-sliced.design/examples)

**Comment by azinit** [comment]
@niyazm524 глянь тоже пож проект в целом)
Мб еще за что глаз зацепится

https://github.dev/EliseyMartynov/fs-vue

**Comment by niyazm524** [comment]
Хотел написать о Vuex, про который Илья выразился "как в эффекторе"

Учитывая, что в Vue 3 reactivity api вынесли в независимый модуль, стало возможным писать сторы, обходясь только `reactive`, `computed` и методами. Если плагины для Vuex не нужны, а отладка не страшное слово, то я вижу это вполне себе выходом из такого мозаичного кода.

Конечно, это всё вкусовщина, но я, признаться, был разочарован, когда они ушли с options API к Composition, не сделав такой ход в Vuex (понятно, что совместимость, но всё же)

Приведу несколько ссылок, описывающих этот метод [раз](https://dev.to/vuesomedev/you-might-not-need-vuex-with-vue-3-52e4) [два](https://vuejsdevelopers.com/2020/10/05/composition-api-vuex/)

**Reply by niyazm524** [reply]
Если в объективщину, то в реалиях стандартного Vuex очень хорошо написано

**Reply by niyazm524** [reply]
Но ещё один плюс в сторону global reactive state - почти нет боли с типами. В Vuex конечно многое уже решено, но вызов методов передачей их названия в строке до сих пор не даёт покоя, и местами даже сам state муторно типизировать.

**Reply by azinit** [reply]
@niyazm524 Спасибо за экспертное ревью! 🚀

**Reply by EliseyMartynov** [reply]
@niyazm524 Мне @azinit сказал, что ты знаешь за бэсикс React'а. Не кажется ли, что подход с провайд инжектами очень плохо будет скейлится? Все-таки Feature Sliced подразумевает, что мы будем писать долгоживущий проект (а значит немаленький). Как аналогию хотел бы привести:

**по ощущениям это как приложение на редакс тулките перенести на контексты (концепции похожи оч сильно у vuex и rtk && провайд/инжекты и context)**

И мое мнение в этом плане, что если мы будем писать средний+ проект на контекстах (рассуждаю как Реактер сейчас), то это очень страшно даже представить)
Просто насколько я слышал от ребят из Vue чатика (давным давно еще, когда только появлялся провайд/инжект) - это именно очень похоже с концептом React Context.

**Reply by niyazm524** [reply]
Да, ты прав, provide/inject это по сути тот же useContext

Возможно, не полностью понял вопрос, но в любом случае, какая бы архитектура не была, часто прибегать к использованию provide/inject = похоронить проект заживо. 
Я считаю, что их использование не должно выходить за рамки разработки плагинов и хуков для них, то есть, чтобы provide был только на root уровне, за исключением случаев, когда ты полностью осознаешь, что делаешь.

**Reply by EliseyMartynov** [reply]
@niyazm524 Понял. Из-за недостаточности времени не могу глубоко вникнуть в паттерн без Vuex. 
Когда мы выбираем Feature Sliced архитектуру, значит подразумеваем достаточно большой проект => исходя из моего поверхностного представления, довольно страшно писать без дебаг тулинга. А как ты выше и написал, с отладкой есть вопросы при таком подходе.

**Comment by EliseyMartynov** [comment]
@niyazm524 еще поделюсь небольшой болью, с которой столкнулся, когда писал на Вью. В Реакте мы можем передать пропом все, что угодно - даже компонент (или даже render-функцию, что почти одно и то же). Во Вью так делать нельзя, если говорим про классику SFC - и если хотим передать компонент в дочерний, то нам нужно делать слот (на сколько я понимаю, если захотим вывернуться с кастомными рендер функциями, то может так и можно сделать, но выглядит это ужасно).
*The problem*:
- вот у нас есть Ant в примере, и у него есть слот actions в компоненте Card. Этот Card у меня непосредственно обернут в entitie, и чтобы в entitie мне передать слот actions из уровня page, мне в моей entitie дополнительно нужно создавать слот actions.
- таким образом я передаю слот actions из уровня page, и описываю его еще на уровне entitie.
- почему так больно? Решается ли это чем-то?

![image](https://user-images.githubusercontent.com/66368523/143778931-15832ead-38bb-4f68-a4f3-c8a9e5e4e10d.png)
![image](https://user-images.githubusercontent.com/66368523/143778959-e62523ec-ce8e-43ce-9cb8-debabbb91afe.png)

**Reply by niyazm524** [reply]
Да, сам не раз сталкивался с подобным, был даже кейс, когда воспользоваться слотом было невозможно, и пришлось писать компонент-обёртку (props{component,renderSlots,renderProps} и setup возвращающий функцию)
Не без доли горести будет сказано, но нормального и удобного способа передачи и наследования слотов я не нашёл :/

**Comment by kaguya3222** [comment]
@niyazm524 @EliseyMartynov Привет) Я из fs community, интересуюсь связкой vue + fs. 

У меня вопрос касательно концепции отделения БЛ от ui логики. Большинство проектов на вью (да и не только вью) миксуют ее прямо в компонентах. Какие у нас есть в целом инструменты для отделения? Как я понял в этом проекте как раз таки vuex и его модули идут как контейнеры для бизнес логики.

Меня немного смущает идея использования стейт менеджера как контейнера для бизнес логики, т.к тот же vuex пораждает глобальный стейт. Не странно ли делать глобальный стейт для какой нибудь формы, которая всего на одной странице используется? Условно говоря, мы уйдем с этой страницы, а данные этой формы во vuex останутся лежать, нам нужно постоянно следить за тем что бы сбрасывать состояние модуля во vuex. Не проблема ли это?

**Reply by niyazm524** [reply]
Да, меня тоже смущает, и не немного. Может быть кейс, что одна и та же форма используется на двух keep-alive вкладках, и тогда их данные становятся общими. Некоторые компоненты хранят очень много данных, а учитывая, что они реактивные, использование RAM увеличивается ещё больше, и хотелось бы чтобы эти данные сами по себе умирали.
Лично я на практике придерживаюсь подхода неких компонентов-абстракций, чтоб на более верхних уровнях писать логику БЛ, но это не лучший вариант, поскольку иногда всё равно приходится миксовать.
Может быть, выносить её в те же хуки..
@azinit, что думаешь?

**Reply by kaguya3222** [reply]
@niyazm524 Идея с хуками звучит хорошо, но вот для меня например это проблема, т.к проект на **Nuxt2**, без возможности миграции на Vue 3, т.к по словам кор тимы, ssr для vue 3, с которым пойдешь уверенно в прод - не готов. И мне бы хотелось юзать подход, который для vue 2 подойдет.

Я вижу такие варианты:
1) ~~миксины~~ (худший наверное)
2) vuex (имеет те проблемы, которые обсудили выше, но хорошо подходит для shared стейта)
3) Взять что то по типу: https://github.com/vuejs/composition-api, в случае, если хуки все же лучший вариант.
4) Renderless компоненты, как компоненты-контейнеры которые поставляют модель данных локального стейта? в целях разделении ui логики и бизнес логики.

Что бы было понятно о чем я говорю, локальный стейт для меня - стейт, который пропадает, как только компонент, использующий его, демонтируется. 

То есть мы можем юзать vuex/pinia/любой другой СТМ для описания модели, которая подходит под shared state, например: `entities/user/model`, но ту же модель данных для локального стейта можно определять внутри renderless компонента:

`**/model/index.vue`
```
export default {
   computed: {
     isLoading() { ... } ,
   },
  methods: {
     doSomeLogic() { ... }
  },
  render() {
     return this.$scopedSlots.default({
        doSomeLogic,
        isLoading
     })
  }
}
```

`**/ui/index.vue`
```
<template>
  <Model v-slot="{ doSomeLogic, isLoading }">
     <span v-if="isLoading>I'm loading</span>
     <button @click="doSomeLogic">Click</button>
  </Model>
</template>
```

Renderless компоненты в принципе являются хорошим решением, когда у тебя есть разные представления, но одна логика. Я бы предложил их использовать для кейсов, когда есть слайс, в котором есть какая то БЛ, но она вообще никак не подходит под shared стейт, но миксовать ее с ui логикой в одном компоненте не хочется.

**Reply by niyazm524** [reply]
Хах, да, согласен, миксины это только если тебе не заплатили. Renderless компоненты действительно выглядят очень выигрышно на фоне всего остального, я использовал этот подход в достаточно редких случаях, так что не могу сказать, что там нет никаких проблем.
Если всё нормально, то даже через метод исключения можно сказать, что это лучший вариант. 
Плагин хуков выглядит, как очень грамотный, но всё же костыль, а создавать инстанс модуля vuex и привязывать его к жизненному циклу компонента (видел подобное) звучит как дичь, да и время онбординга станет 📈

**Reply by kaguya3222** [reply]
@niyazm524 Спасибо за фидбек по поводу renderless компонентов как контейнеров с логикой. 

Продолжу тут тему про стейт менеджеры и их роль как контейнеров для бизнес логики в методологии. Я уже ни раз говорил, и тут, и в группе, и в личке у @azinit, что мне не нравится описывать БЛ каждого модуля в стейт менеджере. Возможно, потому что я за свой маленький опыт использовал только стейт менеджеры, которые реализуют FLUX архитектуру (Vuex например), и они мне всегда казались местом, куда должен попадать только **Application state** (далее - **AS**) . То есть условный список книг, который есть только на одной странице, не должен попадать в **AS**. И в последние дни я как раз изучаю тему с управлением состояния, и я наткнулся на интересное [видео](https://www.youtube.com/watch?v=O5Ijg_wxEqU) Ильи Климова. В нем как раз обсуждаются противоречивые подходы в работе с СТМ flux-like архитектуры, или же просто "как не надо делать". И как раз на [этом](https://youtu.be/O5Ijg_wxEqU?t=358) тайм коде начинается речь про это (но желательно с нуля конечно глянуть). И я в целом с Ильей согласен.

Я двумя руками за то, что бы выносить модель какого-нибудь 'entities/user` в СТМ, т.к это чистый **AS**. Но модель списка книг выносить в стору, когда она используется только в одном месте - мне понять сложно такое.

Поэтому и меня интересуют подходы, как можно отделить БЛ от UI логики, не записывая БЛ в СТМ.

**Reply by azinit** [reply]
> Поэтому и меня интересуют подходы, как можно отделить БЛ от UI логики, не записывая БЛ в СТМ.

Советую как раз прочитать нашумевшую статью про отделение UI и логики от @Kelin2025
(там пример на реакт, но общая суть применима и для Vue по идее)

UPD: Думаю, можно это у самого Ильи подробнее в чате спросить в контексте Вью)
(даже просто вне контекста FSD)

https://kelin2code-mu.vercel.app/posts/stop-writing-logic-in-components

**Reply by azinit** [reply]
> БЛ, но это не лучший вариант, поскольку иногда всё равно приходится миксовать.
> Может быть, выносить её в те же хуки..
> @azinit, что думаешь?

**Мы с @niyazm524 тоже не раз обсуждали как лучше отделять это в мире Вью - и увы там не так много способов, особенно если речь не про Vue3**

Тут кажется действительно почти единственное что остается делать (особенно в контексте FSD) - держать компоненты максимально тупыми (особенно если речь про shared/entities)
А всю логику выстраивать в верхних слоях (даже если придется порой от этого марать компоненты виджетов/страниц)

СТМ тут лучше поможет, но опять же - проблема бойлерплейта и переиспользования от этого не сильно меньше станет
